language/evm/move-to-yul/src/functions.rs

```diff
@@ -107,48 +108,260 @@ impl<'a> FunctionGenerator<'a> {
+                        loop_body: loop_scg_vec,
+                    });
                 } else {
-                    panic!("effective entry block is not basic")
+                    Self::push_non_loop_scg(&mut scg_vec, *blk_id, &cfg, &mut visited_blocks, code);
                 }
-            } else {
-                emitln!(ctx.writer, "let $block := {}", entry_bb);
-                emit!(ctx.writer, "for {} true {} ");
+            }
+        }
+
+        for scg in scg_vec {
+            self.emit_scg(&scg, &cfg, code, ctx, fun_id, target);
+        }
+    }
+
+    pub fn emit_scg(
+        &mut self,
+        scg: &StacklessSCG,
+        cfg: &StacklessControlFlowGraph,
+        code: &[Bytecode],
+        ctx: &Context,
+        fun_id: &QualifiedInstId<FunId>,
+        target: &FunctionTarget,
+    ) {
+        let label_map = Self::compute_label_map(cfg, code);
+        match scg {
+            StacklessSCG::BasicBlock {
+                start_offset,
+                end_offset,
+            } => {
                 ctx.emit_block(|| {
-                    emitln!(ctx.writer, "switch $block");
-                    for blk_id in &cfg.blocks() {
-                        if let BlockContent::Basic { lower, upper } = cfg.content(*blk_id) {
-                            // Emit code for this basic block.
-                            emit!(ctx.writer, "case {} ", blk_id);
-                            ctx.emit_block(|| {
-                                for offs in *lower..*upper + 1 {
-                                    self.bytecode(
-                                        ctx,
-                                        fun_id,
-                                        target,
-                                        &label_map,
-                                        &code[offs as usize],
-                                        true,
-                                    );
-                                }
-                            })
+                    for offs in *start_offset..*end_offset + 1 {
+                        self.bytecode(
+                            ctx,
+                            fun_id,
+                            target,
+                            &label_map,
+                            &code[offs as usize],
+                            false,
+                            true,
+                        );
+                    }
+                });
+            }
+            StacklessSCG::IfBlock {
+                cond,
+                if_true,
+                if_false,
+            } => {
+                emitln!(ctx.writer, "switch $t{} ", cond);
+                emit!(ctx.writer, "case 0 ");
+                if let StacklessSCG::BasicBlock {
+                    start_offset,
+                    end_offset,
+                } = **if_false
+                {
+                    ctx.emit_block(|| {
+                        for offs in start_offset..end_offset + 1 {
```

Comment:
Something not right here. You are already emitting code for a BasicBlock on line 242. Why do you need to do it again here? And as said, what happens if `if_true` and `if_false` are not basic blocks? This should be a recursive algorithm here which handles the variants of StacklessSCG exactly once.